### ISSUE TITLE
Weird directory structure on staging servers

### ISSUE THREAD

**Issue by madclam** [codalab/codalab-competitions#1981]
Here is the directory structure we used to have in the directory in which PARTICIPANT CODE is executed on competitions.codalab.org:
/codalabtemp/tmp3tsanp/run/program
/codalabtemp/tmp3tsanp/run/program/README.ipynb
/codalabtemp/tmp3tsanp/run/program/run.py
/codalabtemp/tmp3tsanp/run/program/public_data
/codalabtemp/tmp3tsanp/run/program/__MACOSX
/codalabtemp/tmp3tsanp/run/program/metadata
/codalabtemp/tmp3tsanp/run/program/res
/codalabtemp/tmp3tsanp/run/program/sample_code
/codalabtemp/tmp3tsanp/run/program/public_data/iris_train.data
/codalabtemp/tmp3tsanp/run/program/public_data/iris_valid.data
/codalabtemp/tmp3tsanp/run/program/public_data/iris_label.name
/codalabtemp/tmp3tsanp/run/program/public_data/iris_feat.name
/codalabtemp/tmp3tsanp/run/program/public_data/iris_test.data
/codalabtemp/tmp3tsanp/run/program/public_data/iris_train.solution
/codalabtemp/tmp3tsanp/run/program/__MACOSX/public_data
/codalabtemp/tmp3tsanp/run/program/__MACOSX/res
/codalabtemp/tmp3tsanp/run/program/__MACOSX/sample_code
/codalabtemp/tmp3tsanp/run/program/res/iris_test.predict
/codalabtemp/tmp3tsanp/run/program/res/iris_valid.predict
/codalabtemp/tmp3tsanp/run/program/sample_code/data_manager.py
/codalabtemp/tmp3tsanp/run/program/sample_code/libscores.py
/codalabtemp/tmp3tsanp/run/program/sample_code/classifier.py
/codalabtemp/tmp3tsanp/run/program/sample_code/data_io.py
/codalabtemp/tmp3tsanp/run/program/sample_code/classifier.pyc
/codalabtemp/tmp3tsanp/run/program/sample_code/libscores.pyc
/codalabtemp/tmp3tsanp/run/program/sample_code/data_io.pyc
/codalabtemp/tmp3tsanp/run/program/sample_code/data_manager.pyc
/codalabtemp/tmp3tsanp/run/program/sample_code/data_converter.py
/codalabtemp/tmp3tsanp/run/program/sample_code/data_converter.pyc

We have something similar on competition-test:
/codalabtemp/tmpvjzPGe/run/program
/codalabtemp/tmpvjzPGe/run/program/sample_code
/codalabtemp/tmpvjzPGe/run/program/metadata
/codalabtemp/tmpvjzPGe/run/program/public_data
/codalabtemp/tmpvjzPGe/run/program/README.ipynb
/codalabtemp/tmpvjzPGe/run/program/res
/codalabtemp/tmpvjzPGe/run/program/run.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/libscores.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_io.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_io.pyc
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_manager.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_converter.pyc
/codalabtemp/tmpvjzPGe/run/program/sample_code/classifier.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/classifier.pyc
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_converter.py
/codalabtemp/tmpvjzPGe/run/program/sample_code/libscores.pyc
/codalabtemp/tmpvjzPGe/run/program/sample_code/data_manager.pyc
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_test.data
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_train.data
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_label.name
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_train.solution
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_feat.name
/codalabtemp/tmpvjzPGe/run/program/public_data/iris_valid.data

BUT, on the new staging server, we have this weird stuff, it looks like we are executing code in the root directory???

/
//opt
//tmp
//etc
//dev
//run
//lib64
...
//sys/module/nf_conntrack_ipv4/initsize

**Comment by ckcollab** [comment]
Those are the files in the docker container, not the host machine, I believe. What code are you running to produce this output? The current working directory is `/`? That doesn't seem right.. but this may also depend on the docker image you are using to run your submission?

**Comment by madclam** [comment]
That's the first time I see that. I was running test before on several servers using queues and dockers (see.4c, docker-server) and I checked on https://competitions-test.codalab.org/, which seems to ALSO have queues. The directory structure was the same as in Codalab v1. The code is run in a directory respecting the structure of the code submission.

**Comment by ckcollab** [comment]
It looks like this is happening because our "docker in docker" container just leaves the current working dir in root?

https://github.com/codalab/docker-codalab-legacy-worker/blob/master/Dockerfile

vs

https://github.com/codalab/codalab-competitions/blob/develop/docker/worker_compute/Dockerfile

(later contains `WORKDIR`)

I believe the fix would be to set a proper `WORKDIR` in that docker image? Will test soon!

**Comment by madclam** [comment]
I managed to find the problem by comparing runs on https://codalab.lri.fr/ and http://13.82.145.119/. The LRI server runs docker and queues, but an older version of them. Something has changed (maybe because of the ingestion program?).
On the LRI server, we have:
-- Current directory /codalabtemp/tmpjjSfmX/run: this is where the program is run.
which includes as subdirectories:
- program: this is where the code is found (including libraries in a directory structure that was submitted in the zip file of code submission)
- input: the input data

On the staging server, we have:
-- Current directory /: this is there the program is run.
- input, output, and program directories are found in: /tmp/codalab/tmpPuRinv/run

I was used to the code running in the "run/" directory, this is why I had problems.

Do we want to run the code from the root directory? Does this not create security issues?

**Comment by madclam** [comment]
I have now conducted extensive testing of the ingestion program feature http://codalab-competitions.readthedocs.io/en/latest/Creating%20a%20Competition%20Bundle/ on both staging servers:
http://13.82.145.119/
and
http://13.90.249.6/

I use the test competition http://13.82.145.119/competitions/55. Eric is an admin on it.
The ingestion program is http://13.82.145.119/my/datasets/download/b64811de-57a3-45e1-9d4e-4988344df793
The sample submission is http://13.82.145.119/my/competition/submission/537/input.zip

Output from ingestion:
http://13.82.145.119/my/competition/submission/537/ingestion_program_stderr_file.txt
Output from prediction:
http://13.82.145.119/my/competition/submission/537/predict_stdout.txt

REMARKS:
- The hidden directory is empty (as viewed by the ingestion program). It should contain the reference data. It is not visible from the participant's code (which is great).
- The submission made by the participant (result or code) cannot be seen from the ingestion program (which is bad). To make is available, the participant's submission has to copy it in the shared directory. This works if we put some special code to that effect inside the python script. But we rather avoid it. Putting an instruction in the metadata file:
command: cp -R $program/ $shared/ (Unix)
command: copy $program\* $shared (Windows)
does not work (why?)

It would be best to make the participant bundle available to the ingestion program. See: https://github.com/codalab/codalab-competitions/issues/1987

**Comment by ckcollab** [comment]
If you run the same submission again, the current working directory should be the run directory...! 

```
*** SCORING PROGRAM: PLATFORM SPECIFICATIONS ***


=== SYSTEM ===


    dist: ('debian', '7.4', '')
    linux_distribution: ('debian', '7.4', '')
    system: Linux
    machine: x86_64
    platform: Linux-4.4.0-92-generic-x86_64-with-debian-7.4
    uname: ('Linux', 'ff50a0b48427', '4.4.0-92-generic', '#115-Ubuntu SMP Thu Aug 10 09:04:33 UTC 2017', 'x86_64', '')
    version: #115-Ubuntu SMP Thu Aug 10 09:04:33 UTC 2017
    mac_ver: ('', ('', '', ''), '')
    memory: svmem(total=16809123840L, available=13387558912L, percent=20.4, used=13708546048L, free=3100577792L, active=6069530624, inactive=6040662016, buffers=553713664L, cached=9733267456)
    number of CPU: 4
    
=== DIRECTORIES ===

-- Current directory /tmp/codalab/tmplwIy5q/run:
.
./ingestion_stderr_file.txt
./ingestion_stdout_file.txt
./input
./metadata
./output
./private_output
./program
./stderr.txt
./stdout.txt
./temp
./tmp5Ra468
./tmpTOhDAI.zip
./tmp_U4i2q
./tmpzFcGQ7.txt
./input/coopetition
./input/history
./input/metadata
./input/ref
./input/res
./input/scores
./input/tmp0WdEtO.zip
./input/tmp3i8Rr8.zip
./input/tmpDUbyWx.txt
./input/tmpHFP5ev.zip
./input/tmphwcPWI.txt
./output/metadata
./output/scores.html
./output/scores.txt
./private_output/metadata
./program/__MACOSX
./program/libscores.py
./program/libscores.pyc
./program/metadata
./program/metric.txt
./program/score.py

-- Input directory /tmp/codalab/tmplwIy5q/run/input:
/tmp/codalab/tmplwIy5q/run/input
/tmp/codalab/tmplwIy5q/run/input/coopetition
/tmp/codalab/tmplwIy5q/run/input/history
/tmp/codalab/tmplwIy5q/run/input/metadata
/tmp/codalab/tmplwIy5q/run/input/ref
/tmp/codalab/tmplwIy5q/run/input/res
/tmp/codalab/tmplwIy5q/run/input/scores
/tmp/codalab/tmplwIy5q/run/input/tmp0WdEtO.zip
/tmp/codalab/tmplwIy5q/run/input/tmp3i8Rr8.zip
/tmp/codalab/tmplwIy5q/run/input/tmpDUbyWx.txt
/tmp/codalab/tmplwIy5q/run/input/tmpHFP5ev.zip
/tmp/codalab/tmplwIy5q/run/input/tmphwcPWI.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/coopetition_downloads.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/coopetition_phase_1.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/coopetition_phase_2.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/coopetition_scores_phase_1.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/coopetition_scores_phase_2.txt
/tmp/codalab/tmplwIy5q/run/input/coopetition/current_user.txt
/tmp/codalab/tmplwIy5q/run/input/history/metadata
/tmp/codalab/tmplwIy5q/run/input/ref/__MACOSX
/tmp/codalab/tmplwIy5q/run/input/ref/iris_valid.solution
/tmp/codalab/tmplwIy5q/run/input/res/answer_from_ingestion.txt
/tmp/codalab/tmplwIy5q/run/input/res/answer_from_submission.txt
/tmp/codalab/tmplwIy5q/run/input/res/metadata
/tmp/codalab/tmplwIy5q/run/input/scores/metadata

-- Output directory  /tmp/codalab/tmplwIy5q/run/output:
/tmp/codalab/tmplwIy5q/run/output
/tmp/codalab/tmplwIy5q/run/output/metadata
/tmp/codalab/tmplwIy5q/run/output/scores.html
/tmp/codalab/tmplwIy5q/run/output/scores.txt


=== METADATA ===

-- Current directory /tmp/codalab/tmplwIy5q/run:
none
-- Input directory /tmp/codalab/tmplwIy5q/run/input:
none

=== VERSIONS ===

Scoring program version: 1.5

Python version: 2.7.11 |Anaconda 2.5.0 (64-bit)| (default, Dec  6 2015, 18:08:32) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

Versions of libraries installed:
abstract-rendering==0.5.1
alabaster==0.7.7
anaconda-client==1.2.2
argcomplete==1.0.0
astropy==1.1.1
babel==2.2.0
backports-abc==0.4
backports.ssl-match-hostname==3.4.0.2
beautifulsoup4==4.4.1
bitarray==0.8.1
blaze==0.9.0
bokeh==0.11.0
boto==2.39.0
bottleneck==1.0.0
cdecimal==2.3
cffi==1.2.1
clyent==1.2.0
colorama==0.3.6
conda-build==1.19.0
conda-env==2.4.5
conda==3.19.1
configobj==5.0.6
cryptography==0.9.1
cycler==0.9.0
cython==0.23.4
cytoolz==0.7.5
datashape==0.5.0
decorator==4.0.6
docutils==0.12
dynd==f641248
enum34==1.1.2
et-xmlfile==1.0.1
fastcache==1.0.2
flask==0.10.1
funcsigs==0.4
futures==3.0.3
gevent-websocket==0.9.5
gevent==1.0.2
greenlet==0.4.9
grin==1.2.1
h5py==2.5.0
idna==2.0
ipaddress==1.0.14
ipykernel==4.2.2
ipython-genutils==0.1.0
ipython==4.0.3
ipywidgets==4.1.1
itsdangerous==0.24
jdcal==1.2
jedi==0.9.0
jinja2==2.8
jsonschema==2.4.0
jupyter-client==4.1.1
jupyter-console==4.1.0
jupyter-core==4.0.6
jupyter==1.0.0
llvmlite==0.8.0
lxml==3.5.0
markupsafe==0.23
matplotlib==1.5.1
mistune==0.7.1
multipledispatch==0.4.8
nbconvert==4.1.0
nbformat==4.0.1
networkx==1.11
nltk==3.1
nose==1.3.7
notebook==4.1.0
numba==0.23.1
numexpr==2.4.6
numpy==1.10.4
odo==0.4.0
openpyxl==2.3.2
pandas==0.17.1
path.py==0.0.0
patsy==0.4.0
pep8==1.7.0
pexpect==3.3
pickleshare==0.5
pillow==3.1.0
pip==8.0.2
ply==3.8
psutil==3.4.2
ptyprocess==0.5
py==1.4.31
pyasn1==0.1.9
pycairo==1.10.0
pycosat==0.6.1
pycparser==2.14
pycrypto==2.6.1
pycurl==7.19.5.3
pyflakes==1.0.0
pygments==2.1
pyopenssl==0.15.1
pyparsing==2.0.3
pytest==2.8.5
python-dateutil==2.4.2
pytz==2015.7
pyyaml==3.11
pyzmq==15.2.0
qtconsole==4.1.1
redis==2.10.3
requests==2.9.1
rope==0.9.4
scikit-image==0.11.3
scikit-learn==0.17
scipy==0.17.0
setuptools==19.6.2
simplegeneric==0.8.1
singledispatch==3.4.0.3
six==1.10.0
snowballstemmer==1.2.1
sockjs-tornado==1.0.1
sphinx-rtd-theme==0.1.9
sphinx==1.3.5
spyder==2.3.8
sqlalchemy==1.0.11
statsmodels==0.6.1
sympy==0.7.6.1
tables==3.2.2
terminado==0.5
toolz==0.7.4
tornado==4.3
traitlets==4.1.0
unicodecsv==0.14.1
werkzeug==0.11.3
wheel==0.26.0
xlrd==0.9.4
xlsxwriter==0.8.4
xlwt==1.0.0```

**Comment by ckcollab** [comment]
> The submission made by the participant (result or code) cannot be seen from the ingestion program (which is bad). To make is available, the participant's submission has to copy it in the shared directory. This works if we put some special code to that effect inside the python script. But we rather avoid it. Putting an instruction in the metadata file:
command: cp -R $program/ $shared/ (Unix)
command: copy $program* $shared (Windows)
does not work (why?)

Yeah, that should be working.. I'll have to dig into this more. I'll make a new issue for this.
